tuf/tuf.go

```diff
@@ -173,23 +162,133 @@ func (tr *Repo) RemoveBaseKeys(role string, keyIDs ...string) error {
+		pubKey, ok := tr.Root.Signed.Keys[keyID]
+		if !ok {
+			return data.BaseRole{}, fmt.Errorf("key with ID %s, specified as a signing key for role %s, was not found in root metadata", keyID, name)
+		}
+		pubKeys[keyID] = pubKey
+	}
+
+	return data.BaseRole{
+		Name:      name,
+		Keys:      pubKeys,
+		Threshold: roleData.Threshold,
+	}, nil
+}
+
+// GetDelegationRole gets a delegation role from this repo's metadata, walking from the targets role down to the delegation itself
+func (tr *Repo) GetDelegationRole(name string) (data.DelegationRole, error) {
+	if !data.IsDelegation(name) {
+		return data.DelegationRole{}, data.ErrInvalidRole{Role: name, Reason: "invalid delegation name"}
+	}
+	if tr.Root == nil {
+		return data.DelegationRole{}, ErrNotLoaded{data.CanonicalRootRole}
+	}
+	_, ok := tr.Root.Signed.Roles[data.CanonicalTargetsRole]
+	if !ok {
+		return data.DelegationRole{}, ErrNotLoaded{data.CanonicalTargetsRole}
+	}
+	// Traverse target metadata, down to delegation itself
+	// Get all public keys for the base role from TUF metadata
+	signedTargetData, ok := tr.Targets[data.CanonicalTargetsRole]
+	if !ok {
+		return data.DelegationRole{}, ErrNotLoaded{data.CanonicalTargetsRole}
+	}
+
+	// Start with top level roles in targets
+	delegationRoles := signedTargetData.Signed.Delegations.Roles
+	var foundRole *data.Role
+	for len(delegationRoles) > 0 {
+		delgRole := delegationRoles[0]
+		delegationRoles = delegationRoles[1:]
+
+		// If this role is delegated above or is our desired role, traverse it
+		if delgRole.Name == name || strings.HasPrefix(name, delgRole.Name+"/") {
+
+			// If we found the role, we can exit the loop
+			if delgRole.Name == name {
+				foundRole = delgRole
+				break
+			}
+
+			// If this is a parent role, keep traversing
+			if delegationMeta, ok := tr.Targets[delgRole.Name]; ok {
+				delegationRoles = append(delegationRoles, delegationMeta.Signed.Delegations.Roles...)
+			}
+		}
+	}
+
+	if foundRole == nil {
+		return data.DelegationRole{}, data.ErrInvalidRole{Role: name, Reason: "delegation does not exist"}
+	}
+
+	pubKeys := make(map[string]data.PublicKey)
+	parentData, ok := tr.Targets[path.Dir(foundRole.Name)]
+	if !ok {
+		return data.DelegationRole{}, ErrNotLoaded{path.Dir(path.Dir(foundRole.Name))}
+	}
+	for _, keyID := range foundRole.KeyIDs {
+		pubKey, ok := parentData.Signed.Delegations.Keys[keyID]
+		if !ok {
+			return data.DelegationRole{}, ErrNotLoaded{name}
+		}
+		pubKeys[keyID] = pubKey
+	}
+
+	return data.DelegationRole{
+		BaseRole: data.BaseRole{
+			Name:      name,
+			Keys:      pubKeys,
+			Threshold: foundRole.Threshold,
+		},
+		Paths: foundRole.Paths,
+	}, nil
+}
+
 // GetAllLoadedRoles returns a list of all role entries loaded in this TUF repo, could be empty
 func (tr *Repo) GetAllLoadedRoles() []*data.Role {
-	return tr.keysDB.GetAllRoles()
+	var res []*data.Role
+	for name, rr := range tr.Root.Signed.Roles {
+		res = append(res, &data.Role{
+			RootRole: *rr,
```

Comment:
I'm getting confused as to what `RootRole` means. It seems like `RootRole` can have any of the 4 base roles.